Firestore/core/src/firebase/firestore/api/firestore.mm

```diff
@@ -143,9 +144,23 @@
   // The client must be initialized to ensure that all subsequent API usage
   // throws an exception.
   EnsureClientConfigured();
+  client_running_ = false;
   [client_ shutdownWithCallback:std::move(callback)];
 }
 
+void Firestore::ClearPersistence(util::StatusCallback callback) {
+  if (client_running_) {
```

Comment:
This isn't thread safe.

We already use `mutex_` in this class to guard whether not the client is configured so we should reuse that here. See `EnsureClientConfigured` for how to acquire and hold the mutex. This works because of C++ scope rules. The lock guard's destructor releases the lock at the close of the scope.

There's also a race condition here: A client could call:
```objc
FIRFirestore *firestore = ...;
[firestore clearPersistenceWithCompletion:...];
[[firestore documentWithPath:@"foo/bar"] setData:@{@"value": @42}];
```

This would get past your check and probably succeed in configuring the client before `-clearPersistence` finished its thing.

Double-checking inside the enqueued lambda will only corruption due to the race but doesn't completely solve the problem. The issue is that when we double check, we're running on a different thread, and if we find that the client is running, throwing an exception there will no longer cleanly indicate where in the user's code they made the error.

To me this means that maybe we should re-evaluate whether or not this method throws initially. Maybe this should just complete the callback with a failed status instead? This may address Doug's comments in the API discussion doc too.

If we did go down that path, it would mean that you could move this check entirely into the callback.

Reply:
OK, we discussed this in chat and concluded that synchronously reporting the usage error here is too much trouble to be worth it. This means we should make the following changes:

  * No longer check or ThrowIllegalState here in the caller's thread
  * Report errors with a Status with a failed precondition code
  * In the worker queue:
    * acquire the mutex_
    * proceed if the client is still unset (the clear-persistence-on-startup case)
    * proceed if the client is set but is shutdown (the clear-persistence-on-shutdown case)

This means you can get rid of the `client_running_` variable in here and obviates some of the other advice in here about about how to protect its usage.